Wrapper around HotwordDetector class to allow for multithreading with detectors

rohantib opened the pull request:
This provides a simple, straightforward interface to using Snowboy detectors in a separate thread to allow for a responsive UI. The wrapper class implements methods ThreadDetector.start() to start the thread, ThreadedDetector.start_recog() to begin hotword detection, ThreadedDetector.stop_recog() to halt detection, and ThreadedDetector.terminate() to terminate the detector and end the thread. 

To keep things consistent, ThreadedDetector.__init__() accepts keworded arguments to be relayed to HotwordDetector.__init__(), and ThreadedDetector.start_recog() accepts keworded arguments to be relayed to HotwordDetector.start(). With the latter, there is the exception of the interrupt_check argument, which is internally implemented by ThreadedDetector so that ThreadedDetector.stop_recog() can stop the detection. So, instead of implementing interrupt_check themselves, users of the ThreadedDetector class only need to call ThreadedDetector.stop().

Lastly, the ThreadedDetector class implements the method ThreadedDetector.is_running(), which returns a boolean for if the detectors are running or not, and the setters ThreadedDetector.change_models() and ThreadedDetector.change_sensitivity() to change the models and the sensitivity of the detectors, respectively. The re-initialization of the HotwordDetector objects is handled internally.

rohantib commented:
Anything new on this PR?

chenguoguo commented:
In the morning I will write my reviews!

chenguoguo commented on examples/Python/snowboythreaded.py:
```diff
@@ -0,0 +1,94 @@
+        self.run_kwargs = kwargs
+        self.commands.put("Start")
+
+    def stop_recog(self):
```
I'd use `pause_recog` instead.

chenguoguo commented on examples/Python/snowboythreaded.py:
```diff
@@ -0,0 +1,94 @@
+
+    def __init__(self, models, **kwargs):
+        """
+        Initialize Detectors object. **kwargs is for any __init__ keyword arguments to be passed into HotWordDetector
```
Make sure you don't put more than 80 characters in a line. Same for other places.

rohantib commented:
I renamed `stop_recog` to `pause_recog`, trimmed all lines below 80 characters, and added a `demo4.py` to demonstrate it. I also co-developed an application utilizing the threaded detectors, hosted at https://github.com/chrishorton/Hands-Free-Presentation. This is a better example of the application and usefulness of incorporating threading in the detectors, though the demo script also demonstrates it well.

rohantib commented:
Any other changes to make for this PR?

xuchen commented:
Just tried it out, this is great! Merging.
